pkg/bloomcompactor/sharding.go

```diff
@@ -0,0 +1,73 @@
+package bloomcompactor
+
+import (
+	"github.com/grafana/dskit/ring"
+)
+
+var (
+	// TODO: Should we include LEAVING instances in the replication set?
+	RingOp = ring.NewOp([]ring.InstanceState{ring.JOINING, ring.ACTIVE}, nil)
+)
+
+// ShardingStrategy describes whether compactor "owns" given user or job.
+type ShardingStrategy interface {
+	OwnsTenant(tenant string) (bool, error)
+	OwnsJob(job Job) (bool, error)
+}
+
+type ShuffleShardingStrategy struct {
+	ring           *ring.Ring
+	ringLifeCycler *ring.BasicLifecycler
+	limits         Limits
+
+	// Buffers to avoid allocations in ring.Get
+	bufDescs           []ring.InstanceDesc
+	bufHosts, bufZones []string
+}
+
+func NewShuffleShardingStrategy(r *ring.Ring, ringLifecycler *ring.BasicLifecycler, limits Limits) *ShuffleShardingStrategy {
```

Comment:
I don't think we need to reimplement (copy/paste) this code. If we only need to extend functionality for `OwnsJob`, we can try struct-embedding. Is there something I'm missing?

Reply:
Right, we can use the sharding implementation from the gateway, move it to `github.com/grafana/loki/pkg/util/ring` and extend it.

Reply:
Done, I created `pkg/util/ring/sharding.go` with strategies to shard by tenant and by fingerprint. Refactored the sharding strategies from bloom-gw and bloom-compactor to reuse the new ones from the ring utils lib.